services/api/src/lib/nft-storage-adapter.ts

```diff
@@ -0,0 +1,155 @@
+  animation_url_mimetype?: string | undefined
+  animation_url?: string | undefined
+  decimals: number
+  description?: string | undefined
+  external_url_mimetype: string
+  external_url: string
+  image_integrity: string
+  image_mimetype: string | undefined
+  image: string
+  name: string
+}
+
+export interface StoreMediaInput {
+  animationUrl?: string
+  imageUrl: string
+}
+
+export interface StoreMediaOutput {
+  animation_url_integrity?: string | undefined
+  animation_url_mimetype?: string | undefined
+  animation_url?: string | undefined
+  image_integrity: string
+  image_mimetype: string | undefined
+  image: string
+}
+
+export interface StoreMetadataInput extends StoreMediaOutput {
+  description?: string
+  editionNumber: number
+  name: string
+  totalEditions: number
+}
+
+export interface NFTStorageAdapterOptions {
+  pinataApiKey: string
+  pinataApiSecret: string
+}
+export default class NFTStorageAdapter {
+  client: PinataClient
+  logger = logger.child({ context: this.constructor.name })
+
+  constructor(private readonly options: NFTStorageAdapterOptions) {
+    this.client = pinataSDK(options.pinataApiKey, options.pinataApiSecret)
+  }
+
+  hashMetadata(metadata: ARC3Metadata) {
+    //SHA-256 digest of the JSON Metadata file as a 32-byte string
+    const hash = crypto.createHash('sha256')
+    hash.update(JSON.stringify(metadata))
+    return hash.digest('hex')
+  }
+
+  mapToMetadata({
+    editionNumber,
+    description,
+    name,
+    totalEditions,
+    ...rest
+  }: StoreMetadataInput): ARC3Metadata {
+    /**
+     * Construct full metadata based on ARC3 Spec
+     * https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0003.md#json-metadata-file-schema
+     */
+    return {
+      name: `${name} ${editionNumber}/${totalEditions}`,
+      decimals: 0,
+      external_url: Configuration.webUrl,
+      external_url_mimetype: 'text/html',
+      ...(description && { description }),
+      ...rest,
+    }
+  }
+
+  async storeFile(url: string) {
+    try {
+      const pipeline = promisify(stream.pipeline)
+
+      // Kick off download stream, intercept file metadata
+      const downloadStream = got.stream(url)
+      const fileName = downloadStream.options.url.pathname
+        .split('/')
+        .at(-1) as fs.PathLike
+      const { mime, stream: outStream } = await getMimeType(downloadStream)
+      outStream.on('error', (error: Error) => {
+        this.logger.error(error.message, `Failed to download ${fileName}`)
+      })
+
+      // Pipe downloaded file to fs
+      const fileWriteStream = fs
+        .createWriteStream(fileName)
+        .on('error', (error: Error) => {
+          this.logger.error(error.message, `Failed to save ${fileName}`)
+        })
+      await pipeline(outStream, fileWriteStream)
+
+      // Read file and send to IPFS
+      const fileReadStream = fs.createReadStream(fileName)
+      const fileUpload = await this.client.pinFileToIPFS(fileReadStream)
```

Comment:
If we fail at the pinata request do we want to `fs.unlink(fileName)` in the catch?

Reply:
Good call, thank you!